design/open/delta-diff.md

```diff
@@ -0,0 +1,186 @@
+
+#### Plugin system consideration
+
+* The plugin system should be flexible enough such that it won't impose a language restriction, so that in the future we could have lakeFS users take part in plugins creation (not restricted to diffs).
+* The plugin system should support multiple OSs.
+* The plugin system should be easy enough to write plugins.
+
+#### Hashicorp Plugin system
+
+Hashicorp's battle-tested [`go-plugin`](https://github.com/hashicorp/go-plugin) system is a plugin system over RPC (used in Terraform, Vault, Consul, and Packer).
+Currently, it's only designed to work over a local network.  
+Plugins can be written and consumed in almost every major language. This is achieved by supporting [gRPC](http://www.grpc.io/) as the communication protocol.
+* The plugin system works by launching subprocesses and communicating over RPC (both `net/rpc` and `gRPC` are supported).
+  A single connection is made between any plugin and the core process. For gRPC-based plugins, the HTTP2 protocol handles [connection multiplexing](https://freecontent.manning.com/animation-http-1-1-vs-http-2-vs-http-2-with-push/).
+* The plugin and core system are separate processes, which means that a crash of a plugin won't cause the core system to crash (lakeFS in our case).
+
+![Microkernel architecture overview](diagrams/microkernel-overview.png)
+[(excalidraw file)](diagrams/microkernel-overview.excalidraw)
+
+---
+
+## Implementation
+
+![Delta Diff flow](diagrams/delta-diff-flow.png)
+[(excalidraw file)](diagrams/delta-diff-flow.excalidraw)
+
+#### DiffService
+
+The DiffService will be an internal component in lakeFS which will serve as the Core system.  
+In order to realize which diff plugins are available, we shall use new configuration values:
+1. `LAKEFS_PLUGINS_LOCATION`
+   The DiffService will use this location to search for the required plugin. Defaults to lakeFS's configuration file directory.
+2. `LAKEFS_PLUGIN_DIFF_{TYPE OF DIFF}`
+   The DiffService will use the name of binary provided by this env var to perform the diff. For instance, `LAKEFS_PLUGIN_DIFF_DELTA=deltaDiffBinary`
+   The DiffService will use the [plugins' location](https://github.com/hashicorp/terraform/blob/main/plugins.go) and the diff binary to load the plugin and request a diff.
+   The **type** of diff will be sent as part of the request to lakeFS as specified [here](#API).
+   The communication between the DiffService and the plugins, as explained above, will be through the `go-plugin` package (`gRPC`).
+
+#### Delta Diff Plugin
+
+Implemented using [delta-rs](https://github.com/delta-io/delta-rs) (Rust), this plugin will perform the diff operation using table paths provided by the DiffService through a `gRPC` call.  
+To query the Delta Table from lakeFS, the plugin will generate an S3 client (this is a constraint imposed by the `delta-rs` package) and send a request to lakeFS's S3 gateway.  
+The diff algorithm:
```

Comment:
What is the expected amount of calls to the gateway? How much data is to be retrieved from the object store? Two things to consider:
1. IIUC we have a request waiting in the background, we should be fast..
2. How we scale.. If users perform several concurrent requests, will we choke?

Reply:
As we don't really know the amount of Delta Log files used on average (or median, or max) the amount of data is currently an unknown. There will be two calls to the S3 gateway- one history command for each table.
One of the metrics specified below is getting the number of files That Delta users use.
Regarding point 1, for the MVP, this process can be slow (the whole thing is tagged as experimental), but it will help us evaluate the concurrency and resources needed for the Diff.
Regarding point 2, scaling the requests from the Diff service to the diff plugin is pretty easy as it uses an HTTP/2 connection to communicate (we can run multiple streams on a single connection). If a user makes multiple requests for a diff we can implement a backoff in the plugin itself.

Reply:
Not sure I understand, is "history" a gateway command or a delta one? I'm not familiar with that gw endpoint so I'm guessing it's a delta command that executes s3 requests to the gw. I suggest we have a rough estimate for the number and type of calls now. I'm sure we'll be able to handle it but it's better to understand the costs sooner rather than later. 
By scale I meant the number of calls to the gateway. If a single diff performs 100 calls that fetches data, then 10 diffs performance is something to consider. I agree that the service-plugin communication scale is not a concern at this point.

Reply:
Gotcha.
So the **delta** history command is quite an expensive command- it will send a request to fetch each delta log (JSON) file one by one from the gateway. That means that if there are N log entries, then N requests will get sent to the gateway. 
This is surely not the way we would want our diff to eventually work, but at this point (for the case of an MVP, and to verify that a diff is even needed) I don't think we should work towards making it efficient.

Reply:
+1 :+1: to measure the number of calls a "simple" history performs.

Reply:
The [second applicative metric](https://github.com/treeverse/lakeFS/pull/4902/files#diff-e66e2e9fcb8162bd781db293307eed89340003eac894d788a44ce24341f64018R183) will basically do that.